spec/live/companies_integration_spec.rb

```diff
@@ -15,9 +15,16 @@
     company.update! name: "Create Delete Test 2"
     company = Hubspot::Company.find_by_id(company.vid)
 
+
     expect(company["name"]).to eql "Create Delete Test 2"
 
+    Hubspot::Company.batch_update!([{objectId: company.vid, name: 'Batch Update'}])
+    company = Hubspot::Company.find_by_id(company.vid)
```

Comment:
When I run the tests I'm hitting a race condition here where sometimes it succeeds with the updated company name and other times it fails. This might be due to the way HubSpot handles reads and writes and is likely dependent on the speed the tests run.

Rather than adding a sleep or something like that, what do you think about something like this that queries the company until it's last modified date is updated and stops at a maximum of 3 tries?:
```ruby
last_modified = company["hs_lastmodifieddate"]

Hubspot::Company.batch_update!([{objectId: company.vid, name: 'Batch Update'}])

3.times do
  company = Hubspot::Company.find_by_id(company.vid)
  break if company["hs_lastmodifieddate"] != last_modified
end
```

Reply:
Interesting question.

I did try this now, but in my case, the execute time range approx 4.3 sec ~ 9.x sec, even the fastest case I never have this situation you mentioned, I never got any error during 50 times testing.

According to HubSpot's doc, limitation is **10 requests per second. 40,000 requests per day**. maybe it's because `demo` (somebody else also used demo at the moment?). This integration test totally includes only 9 requests as what I count.

Your suggestion is doable, somehow it's weird to me. (Only this company `find_by_id` might or might not be succeed? how about other methods after?)

Maybe we should separate every action and put into different `it` blocks, maybe we could put `retry` into `begin...rescue` block and put all code into it. Either ways, I feel this integrate spec is weird, and put retry 3 times under specific code make it more weird 😬

if this code is meaned to be succeed, it should succeed every time; if under `demo` would cause this time limit error, then we shouldn't have this kind of test spec because it's not reliable. wdyt @cbisnett?

Reply:
So I've confirmed on my side that when the test fails it's because the last modified date returned from the `find_by_id` call hadn't changed. What I think this means is that the bulk update takes a few milliseconds to propagate and I'm likely seeing the original data in the first fetch and the second fetch gets the updated fields. This appears to be confirmed by the last modified date being updated. This is a pretty typical situation for a setup where there is a primary database instance that receives all writes and read replicas that are always slightly behind the primary.

I agree that this is a pretty gross change and the whole test itself is poorly constructed. Ideally each of these calls would be broken out into their own `it` blocks. These are some of the changes I have planned as I'm going to rewrite all of the tests to upgrade from Rspec 2 to Rspec 3. Rather than have a potentially flaky test that sometimes succeeds and othertimes fails, I think it's probably better to add even a short sleep between the batch update and fetching the company. I agree it's gross but I don't see a better option.

Reply:
make sense, added 👍